### PR TITLE
(feat): Add audit trail support #922

### PR DESCRIPTION
**Pull request by bdhobare**
Issue #922 
## Summary
- Added new column `is_active` to the tables `answer`, `review_score` and `guest_registration_answer`. This column is `true` when the entry is the latest and `false` for historical/updated entries.
- Updated repository methods to filter using this new column.
- Added new column `created_on` which tracks when entries were created.
- Updated the PUT api to insert new entries and mark them as active and mark previous entries as not active.
- Updated `Response.answers` to also filter only active answers.

## Tests done

1. `docker-compose run web bash -c 'cd api; nosetests -v app.responses.tests'`
2. `docker-compose run web nosetests`
3. CircleCI 
3. Manual verification using postman.

### CODE REVIEW

**Review by avishkar58 — CHANGES_REQUESTED** [review]
This looks great, thanks @bdhobare ! 
Can you also add a timestamp so that we know when the answers were updated?

File: api/migrations/versions/794506c08e10_.py
Comment (amritpurshotam) [inline api/migrations/versions/794506c08e10_.py:21]: Can we also please update the accompanying `GuestRegistrationAnswer` model in `./api/app/invitedGuest/models.py` and `ReviewScore` model in `./api/app/reviews/models.py` and their accompanying references similar to what you did for `Answer` as well?

**Review by amritpurshotam — CHANGES_REQUESTED** [review]
We do need to make some updates but overall good job @bdhobare, thanks for your contribution!

File: api/migrations/versions/794506c08e10_.py
Comment (bdhobare) [reply api/migrations/versions/794506c08e10_.py:21]: Hi @amritpurshotam , I thought it would be better for me to submit these changes in 2 separate PRs since putting them here might make this PR too big and hard to review. What do you think?

Comment (amritpurshotam) [reply api/migrations/versions/794506c08e10_.py:21]: Hey @bdhobare, that's a good point but since these are quite similar and tightly coupled components I don't feel it would be more difficult on me to review. Perhaps if it helps you to develop against and keep it small on your side, maybe branch off this one, make the changes there, then merge back here once it's ready for review?

Comment (bdhobare) [reply api/migrations/versions/794506c08e10_.py:21]: @amritpurshotam please see the changes.

**Comment by amritpurshotam** [comment]
Hey @bdhobare. Thanks for the updates! Just been a little swamped to review them properly but I should get a chance this weekend. Apologies for the delay.

File: api/migrations/versions/794506c08e10_.py
Comment (amritpurshotam) [inline api/migrations/versions/794506c08e10_.py:28]: None of these columns have been added to the models. Was this intentional? I can see they would default to the current date and the columns and values would exist in the database but I'm uncomfortable with them being _hidden_ in these scripts especially since we're using a model/code first approach. What do you think about moving these to the model class and also explicitly defining them there?

**Review by amritpurshotam — COMMENTED** [review]
Hey @bdhobare. Looks mostly good! Just have a question about the `updated_on` columns, let me know what you think.

File: api/migrations/versions/794506c08e10_.py
Comment (bdhobare) [reply api/migrations/versions/794506c08e10_.py:28]: This definitely makes sense @amritpurshotam . I overlooked defining them in the model class because I saw they are not used in any part of the code because the date is only set and not modified anywhere. Let me declare them in the models too.

Comment (amritpurshotam) [reply api/migrations/versions/794506c08e10_.py:28]: Sorry missed this update 😓 I see they're on the model class now, thanks!
There is one more thing though, I should have mentioned this on the last comment so apologies for the inconvenience.
Can we please rename the column to `created_on` instead and remove the `onupdate=sa.func.now()` bit (in fact I'm in two minds if we should even have the `server_default` and just explicitly set the field manually since that's what's done elsewhere in the codebase)? The reason is that we don't want this value to ever change (especially accidentally) and having the name reflect that would help communicate this rule. 

P.S. you mentioned that the field isn't used anywhere which is a fair assumption. But it will become important later for when we perhaps want to display all the answers (for say an admin of some kind), it would be nice to order them by the creation date and having the field on the model will be required for this. A second reason is that any database changes we make must flow from the codebase (and never the other way round). We have documented the process [here](https://github.com/deep-learning-indaba/Baobab/tree/develop/api#database-migrations) with the commands to run to generate the initial migration. We then edit it if necessary, for eg. with new not null columns, or we edit the code again to make sure the migration we want gets generated correctly). We shouldn't normally be writing migrations from scratch.

Comment (bdhobare) [reply api/migrations/versions/794506c08e10_.py]: Updated with the new recommendations. `server_default` left for now.

**Review by amritpurshotam — APPROVED** [review]

File: api/migrations/versions/794506c08e10_.py
Comment (amritpurshotam) [reply api/migrations/versions/794506c08e10_.py]: @bdhobare Thanks for the update! Approved from my side 😄

**Comment by bdhobare** [comment]
Hi @avishkar58 , waiting for your approval to merge.

**Review by avishkar58 — APPROVED** [review]
Apologies for the delay. Looks good to me, thank you for the contribution! :)

**Pull request merged**
